lib/cpp/solver/asaga.cpp

```diff
@@ -0,0 +1,224 @@
+        next_iterate.mult_incr(iterate, 1.0 / epoch_size);
+      }
+    }
+  };
+
+  std::vector<std::thread> threads;
+  for (size_t i = 0; i < un_threads; i++) {
+    threads.emplace_back(lambda);
+  }
+  for (size_t i = 0; i < un_threads; i++) {
+    threads[i].join();
+  }
+  if (variance_reduction == SAGA_VarianceReductionMethod::Last) {
+    next_iterate = iterate;
+  }
+  TStoSolver<T, std::atomic<T>>::t += epoch_size;
+}
+
+template <class T>
+void AtomicSAGA<T>::solve_sparse_proba_updates(bool use_intercept,
+                                               ulong n_features) {
+  // Data is sparse, and we use the probabilistic update strategy
+  // This means that the model is a child of ModelGeneralizedLinear.
+  // The strategy used here uses non-delayed updates, with corrected
+  // step-sizes using a probabilistic approximation and the
+  // penalization trick: with such a model and prox, we can work only inside the
+  // current support (non-zero values) of the sampled vector of features
+
+  T n_samples = model->get_n_samples();
+  T n_samples_inverse = 1 / n_samples;
+
+  Array<std::atomic<T>> minimizer(model->get_n_coeffs());
+
+  auto lambda = [&](uint16_t n_thread) {
+    T grad_factor_diff = 0;
+    T x_ij = 0;
+    T grad_avg_j = 0;
+    T step_correction = 0;
+    T iterate_j = 0;
+    T grad_i_factor = 0;
+    T grad_i_factor_old = 0;
+
+    struct timespec start, finish;
+    T elapsed;
+#if !defined(_WIN32)  // temporarily disabled TODO DOESN'T WORK ON WINDOWS
+    clock_gettime(CLOCK_MONOTONIC, &start);
+#endif
+    ulong idx_nnz = 0;
+    for (ulong t = 0; t < epoch_size * iterations; ++t) {
+      // Get next sample index
+      ulong i = get_next_i();
+      // Sparse features vector
+      BaseArray<T> x_i = model->get_features(i);
+      grad_i_factor = model->grad_i_factor(i, iterate);
+      grad_i_factor_old = gradients_memory[i].load();
+
+      while (!gradients_memory[i].compare_exchange_weak(grad_i_factor_old,
+                                                        grad_i_factor)) {
+      }
+
+      grad_factor_diff = grad_i_factor - grad_i_factor_old;
+      for (idx_nnz = 0; idx_nnz < x_i.size_sparse(); ++idx_nnz) {
+        // Get the index of the idx-th sparse feature of x_i
+        ulong j = x_i.indices()[idx_nnz];
+        x_ij = x_i.data()[idx_nnz];
+        grad_avg_j = gradients_average[j];
+        // Step-size correction for coordinate j
+        step_correction = steps_correction[j];
+
+        iterate_j = iterate[j].load();
+        // while (!iterate[j].compare_exchange_weak(
+        //   iterate_j,
+        //   iterate_j_t
+        // ));
+
+        while (!gradients_average[j].compare_exchange_weak(
+            grad_avg_j,
+            grad_avg_j + (grad_factor_diff * x_ij * n_samples_inverse))) {
+        }
+
+        // Prox is separable, apply regularization on the current coordinate
+        iterate[j] = casted_prox->call_single(
```

Comment:
Should'nt we use `compare_exchange_weak` here?

Reply:
I don't it's necessary, but feel free to test them